torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -499,6 +507,54 @@ void Reducer::initialize_buckets(
+        // flat contents tensor.
+        //
+        // Gradients may have dense memory but non-row-major-contiguous strides
+        // (e.g. channels_last or channels_last_3d). For coalesced accesses during
+        // copy_s, it's beneficial for each view's layout to match its grad's layout.
+        //
+        // Specifically, we expect torch/csrc/autograd/AccumulateGrad.h produces grads
+        // that obey its "Gradient Layout Contract":
+        //   (1) if variable.is_non_overlapping_and_dense(), the stashed grad's
+        //       strides match variable.
+        //   (2) else, stashed grad is rowmajor contiguous.
+        // and create views to match.
+        //
+        // If AccumulateGrad breaks the contract, and produces a grad with an
+        // unexpected layout, performance will degrade due to poor memory access
+        // patterns when copy_ing grad data in and out of its bucket view.
+        // However, numerics remain correct, because the bucket view is the same on
+        // either end of the raw allreduce.  bucket_view.copy(grad) tranposes (+ densifies)
+        // to the bucket view's layout, the data is allreduced, then grad.copy_(bucket_view)
+        // transposes it back to grad's layout.
+        //
+        // The only way the numerics can go haywire is if the bucket views themselves have
+        // different layouts across processes (or replicas).  Bucket views' sizes and strides are set
+        // based on param layouts, using the same logic that (we expect) AccumulateGrad uses
+        // for their grads.  Therefore, the only way a bucket view could have different layouts
+        // in different processes is if its param has a different layout in different processes.
+        // We can check that param layouts match across processes and replicas in Reducer's
+        // constructor by allreducing some metadata.  Checking just once won't catch if someone
+        // messes with param layouts over time, but not messing with params after DDP construction
+        // is already a documented constraint.
+        for (size_t i = 0; i < replica.variables.size(); i++) {
+          const auto& v = replica.variables[i];
+          if (v.is_non_overlapping_and_dense()) {
+            // If the param's memory is dense, match its layout, anticipating the autograd engine
+            // (AccumulateGrad) will also create gradients matching its layout.
+            replica.bucket_views.push_back(replica.contents
+                                           .as_strided(v.sizes(), v.strides(), replica.offsets[i]));
+          } else {
+            // Fall back to a C-style contiguous view, again anticipating AccumulateGrad will do
+            // the same when stashing grads for non-dense params.
+            replica.bucket_views.push_back(replica.contents
+                                           .as_strided({v.numel()}, {1}, replica.offsets[i])
```

Comment:
I feel like the narrow that was used before here makes it clearer what the intent is.

Reply:
SGTM, i had as_strided on the brain when i was writing that part.